### DISCUSSION TITLE
A replacement for BinaryFormatter in .NET 8

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#92] [General]
# A replacement for BinaryFormatter in .NET 8

In the old .NET Framework days, you could use the BinaryFormatter class to serialize and deserialize objects. This can be convenient for cloning or storing some session states. As the BinaryFormatter has some serious security concerns, the .NET team marked it as obsolete (as error) in .NET 7 and onwards.

https://steven-giesel.com/blogPost/4271d529-5625-4b67-bd59-d121f2d8c8f6

**Comment by VahidN** [comment]
I think this is the equivalent of `BinaryFormatter` using `JsonSerializer`:
```C#
var resultBytes = JsonSerializer.SerializeToUtf8Bytes(data);
var data = JsonSerializer.Deserialize<T>(new ReadOnlySpan<byte>(resultBytes));
```

**Reply by linkdotnet** [reply]
Very good point. I didn't think of that and will add it later to the list!
Thanks @VahidN

**Comment by Ezeji** [comment]
Nice post Steven,

How would you handle the downside of MessagePack not being human-readable in a case where performance is a requirement and having human-readable logs is also necessary which as you pointed out is a strong side of using JSON formatting libraries?

**Reply by linkdotnet** [reply]
Hey @Ezeji,

thanks for the kind words.
I would argue there are multiple ways you could tackle that. One would be putting the logging behind a toggle or very simple, do the logging only in DEBUG mode:

```csharp
#ifdef DEBUG
MessagePack.SerializeToJson(myData); // And log it to a log file
#endif
```

While there can be differences between what MessagePack does when serializing binary or JSON, it can give you a general idea what was going on. 

If there is a hard requirement to log every object even in release/production, you might come around with offloading this via a small background service (or background thread for that matter) that doesn't influence your request thread.

Alternatively, a good testing net might make all of this obsolete or, at the very least, give you much confidence.

So it really depends on your concrete needs.

**Reply by Ezeji** [reply]
Thanks for the prompt response, Steven.

Good thoughts and ideas shared. Will carry out further tests to see various possibilities such as tests related to studying(via benchmark analysis) the memory allocation implications/processing time from serializing binary for processing within the server then deserialiizing into a .NET type and further log/return a JSON serialized data format to a client side in comparison with using say NewtonsoftJson library for serializing JSON, deserializing into a .NET type and further serializing for logging/returning a JSON serialized data format to the client side amongst other possible test scenarios or cases.

**Comment by mizrael** [comment]
great article. Just FYI, I remember a while ago there were some security concerns with MessagePack as well and we were encouraged to not use it on production systems. Not sure it's still the case though...

**Reply by linkdotnet** [reply]
Good catch @mizrael . Indeed there was a CVE: https://github.com/MessagePack-CSharp/MessagePack-CSharp/security/advisories/GHSA-7q36-4xx7-xcxf

It has been fixed in recent versions of the library.

**Comment by wstaelens** [comment]
can Cbor ( https://learn.microsoft.com/en-us/dotnet/api/system.formats.cbor?view=dotnet-plat-ext-8.0 ) be used? does it work with streams?

**Reply by linkdotnet** [reply]
I don't have any prior experience with Cbor.
Reading the description, it might be.

**Comment by chadgi** [comment]
How to handle legacy files which are serialized by binaryformatter? as none of the above serializers will be able to deserialize them.

**Reply by linkdotnet** [reply]
That depends on the given type and context.
Is there a way around serialisation?

**Reply by chadgi** [reply]
Thats the question, am not sure if there is a way, thought I might get some inputs here.
The data which is serialized is a simple class with few properties in it but the problem is, am unable to deserialize it, apart from BinaryFormatter, none (like xml/json/proto-buff etc serializers), for obvious reasons. Unless I go into bit level and extract the data, I don't see any other way to deserialize the file serialized by BinaryFormatter.  Just worried about loosing legacy data.

**Reply by wstaelens** [reply]
facing the same issues for backwards compatibility.

**Reply by linkdotnet** [reply]
If you can’t migrate you can still think of copying the BinaryFormatter from dotnet in your own Codebase.

I would not recommend that, but it could be one measure

**Reply by wstaelens** [reply]
this is what i've seen in other projects also... just what we thought...

**Comment by MagicAndre1981** [comment]
You don't need to copy the code, Microsoft [moved it to a nuget](https://github.com/dotnet/runtime/pull/103255):

https://www.nuget.org/packages/System.Runtime.Serialization.Formatters/9.0.0-preview.6.24327.7

**Reply by linkdotnet** [reply]
Interesting. Thanks for the link. This might be the more sensible choice over copying and pasting the whole thing into your own code.

**Comment by sankethamdapure** [comment]
Great article! I gained a lot of new insights! 
I'm also working on migrating from BinaryFormatter to an alternative solution, but unfortunately, none of the options you mentioned seem to work for me as is. Our situation involves complex objects, and we have constraints that prevent us from adding attributes to classes, as they are used by other systems. Despite the limitations, BinaryFormatter worked well in our scenario. The only viable option I see is writing custom JSON converters, but that feels quite daunting. 
Do you have any recommendations?

**Reply by MagicAndre1981** [reply]
try the mentioned `System.Runtime.Serialization.Formatters` nuget to continue using BinaryFormatter under .net9

**Reply by linkdotnet** [reply]
As @MagicAndre1981 says you can still continue using `BinaryFormatter` for quite a while.

MessagePack and even `Newtonsoft.Json` is pretty forgiving in terms of complex objects (not so much STJ).
The question is, what exactly is the problem? Complex objects could mean deeply nested, but could also mean something else.
Do you have an example where it doesn't work?

**Reply by sankethamdapure** [reply]
Thanks, @MagicAndre1981. I'll give this NuGet package a shot. However, even when using the <EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization> flag, I encounter an error at runtime: "BinaryFormatter serialization and deserialization are not supported on this platform." I'm migrating a legacy Xamarin.iOS project to .NET iOS, so this might be related to some AOT-related restrictions.

Thanks, @linkdotnet. I'm facing a JsonSerializationException when trying to serialize an object that previously worked fine with BinaryFormatter. The challenge lies in our deeply nested complex objects with custom generic types, for instance, System.Collections.Generic.List[CustomKeyValuePair], where CustomKeyValuePair itself contains a list of CustomKeyValuePair for other custom objects.

Yes, I'll try creating custom converters for Newtonsoft.Json, as we already have a few converters in place.

**Comment by tiktoska** [comment]
Hi, I tried MessagePack and must say,was very disappointed, as I tried a simple case with one datetime property.
How could this solution be so popular it does not work?

```csharp
using System.Net.Mime;
using MemoryPack;
using MessagePack;
using MessagePack.Resolvers;

namespace TestProject1
{
    public class CustomException 
    {
        public CustomException()  { }

        private readonly string _text;

        public CustomException( string text)
        {
            this._text = text;
        }

        public DateTime CreationTime { get; set; }
    }

    public class Tests
    {


        [SetUp]
        public void Setup()
        {
        }

        [Test]
        public void Test1()
        {
            CustomException ex = new CustomException("Test");
            ex.CreationTime = DateTime.Now;
            
            var bin = MessagePack.MessagePackSerializer.Serialize(ex, ContractlessStandardResolver.Options);

            var obj = MessagePack.MessagePackSerializer.Deserialize<CustomException>(bin, ContractlessStandardResolver.Options);

            Assert.IsTrue(ex.CreationTime == obj.CreationTime);
            Assert.Pass();
        }
    }
}
```

**Reply by linkdotnet** [reply]
I tried quickly your example and I receive the following:
```csharp
ex.CreationTime
    should be
2024-11-16T12:12:20.6341380Z
    but was
2024-11-16T13:12:20.6341380+01:00
```
Which isn't nice but somewhat understandable. If you would use `ex.CreationTime = DateTime.UtcNow;` then your tests would even pass. That makes sense as `DateTime.Now` has `Local` kind where as `new DateTime(...)` as by default `Unspecified` if not provided (which happens if deserialized).

**Reply by tiktoska** [reply]
Yes thank you. it is a bit strange, but can work. Do you know how can I serialize/deserialize classes without constructors? The main purpose for my evaluation is serializing exceptions which not always are .net exceptions.

**Reply by linkdotnet** [reply]
The absence of a c'tor shouldn't be a problem. Do you have a very minimalistic example?

**Reply by tiktoska** [reply]
Yes, when I remove (in my example) the default parameterless constructor . I get the following error
MessagePack.MessagePackSerializationException : Failed to serialize TestProject1.CustomException value.
  ----> System.TypeInitializationException : The type initializer for 'FormatterCache`1' threw an exception.
  ----> MessagePack.Internal.MessagePackDynamicObjectResolverException : can't find matched
constructor. type:TestProject1.CustomException
   at MessagePack.MessagePackSerializer.Serialize[T](MessagePackWriter& writer, T value, MessagePackSerializerOptions options)
   at MessagePack.MessagePackSerializer.Serialize[T](T value, MessagePackSerializerOptions options, CancellationToken cancellationToken)

**Reply by linkdotnet** [reply]
Thanks.

I guess there are two ways, depending what you like. The "obvious" choice is to make your `Text` property somewhat accessible.
```csharp
public class CustomException 
{ 
    public string Text { get; private set; }

    public DateTime CreationTime { get; set; }
}
```

The other option is to offer "both" c'tor  plus a different (de)serializer:
```csharp
var ex = new CustomException
{
    CreationTime = DateTime.Now
};

var bin = MessagePackSerializer.Serialize(ex, ContractlessStandardResolverAllowPrivate.Options);
var obj = MessagePackSerializer.Deserialize<CustomException>(bin, ContractlessStandardResolverAllowPrivate.Options);

Console.WriteLine(obj.CreationTime);

[MessagePackObject]
public class CustomException
{
    [Key(0)] private readonly string _text;

    [SerializationConstructor]
    private CustomException(string text) => _text = text;
    
    /// So everyone can still instantiate this
    public CustomException()
    {
        _text = "TEST";
    }
    
    [Key(1)]
    public DateTime CreationTime { get; set; }
}
```

**Reply by tiktoska** [reply]
Yes, you are right, with my own classes it is easier. I am thinking about third party classes where I have no control on it. Thanks a lot for your time.

**Reply by linkdotnet** [reply]
Yeah that is a different beast!

**Comment by DomZZ** [comment]
Hello, thanks for this benchmark !

I have a huge graph object that contains about 250K objects that referencing each other.

And I didn't find any replacement for BinaryFormatter fast enough which supports circular references (NewtonSoft.Json is tool slow) ...

By chance, did you have some sucessfull experience on this task ?

Thanks

**Reply by linkdotnet** [reply]
No unfortunately not at all. But, if you can control what is going on, you can still continue using it (of course with strings attached and a potential security risk).
